### PR TITLE
BREAKING CHANGE: support native eth_address by the `ETH Address Registry` layer2 contract

### PR DESCRIPTION
**Pull request by Flouse**
# Interoperability: support native eth_address by the ETH Address Registry layer2 contract

## Feature Update

- feat: add ETH Address Registry layer2 contract
```C
/**
 * `ETH Address Registry` layer2 contract
 * 
 * This contract introduces two-ways mappings between `eth_address` and
 * `gw_script_hash`.
 * 
 *   - `eth_address` is the address of an Ethereum EOA (externally owned account
 *     ) or a Polyjuice contract account.
 * 
 *   - Godwoken account script hash(a.k.a. `gw_script_hash`) is a key used for
 *     locating the account lock. Godwoken enforces one-to-one mapping between 
 *     layer 2 lock script and accountID.
 */
```
  ```C
  // c/eth_addr_reg.c

  /* There are 3 MSG_TYPE in ETH Address Registry layer2 contract  */
  #define MSG_QUERY_GW_BY_ETH 0
  #define MSG_QUERY_ETH_BY_GW 1
  #define MSG_SET_MAPPING     2
  ```

## TODO
- [x] update `polyjuice-tests/src/test_cases`

## Update docs
   - [x] update the docs about comparison with EVM
   - docs: update addition-features.md about transfer_to_any_sudt

### CODE REVIEW

**Review by zeroqn — APPROVED** [review]

File: c/eth_addr_reg.c [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+    return ret;
+  }
+  uint64_t short_script_hash_len = DEFAULT_SHORT_SCRIPT_HASH_LEN;
+  /* sudt */
+  mol_seg_t sudt_id_seg = MolReader_Fee_get_sudt_id(&fee_seg);
+  uint32_t sudt_id = *(uint32_t *)sudt_id_seg.ptr;
+  /* amount */
+  mol_seg_t amount_seg = MolReader_Fee_get_amount(&fee_seg);
+  uint128_t amount = *(uint128_t *)amount_seg.ptr;
+  return sudt_pay_fee(ctx, sudt_id, short_script_hash_len,
```
Comment (jjyr) [inline c/eth_addr_reg.c:43]: This PR https://github.com/nervosnetwork/godwoken-scripts/pull/107 removes the SUDT pay fee feature.
We should do the same here, only support pay CKB as fee.

File: c/polyjuice.mol [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+    eth_address: Byte20,
+}
+
+struct GwToEth {
+    gw_script_hash: Byte32,
+}
+
+struct SetMapping {
+    gw_script_hash: Byte32,
+    fee: Fee,
```
Comment (jjyr) [inline c/polyjuice.mol:21]:
```suggestion
    fee: Uint64, // tx fee in CKB shannon
```

**Review by jjyr — COMMENTED** [review]
1. Please refactoring `fee: Fee` to `fee: Uint64`. since we stopped support SUDT pay fee.
2. EOA | contract to script_hash is using two types of keys, which is unneccesary, to use same type of key we can save one query.
3. This PR assume the first eoa in the whitelist is ETH EOA, which may not be always true.

File: c/eth_addr_reg.c (hunk 1 above)
Comment (Flouse) [reply c/eth_addr_reg.c]: https://github.com/nervosnetwork/godwoken-polyjuice/pull/113/commits/2bedc018e6936d1469413f78c557d492e9407f18

File: Makefile [hunk 3]
```diff
@@ -221,6 +248,14 @@ contract/sudt-erc20-proxy:
 # then echo "ERC20BIN_SHASUM matches" ; \
 # else echo "ERC20BIN_SHASUM does not match" ; exit 1 ; fi
 
+fetch-gw-scripts:
+	mkdir -p build
+	docker pull nervos/godwoken-prebuilds:latest
```
[the remark this answers is not in the document]
Comment (Flouse) [reply Makefile:253]: I will create a new PR after updating the Godwoken and Godwoken-sciprts deps.

**Review by jjyr — COMMENTED** [review]
LGTM, Please fix CI error

**Pull request merged**